File: src/main/java/com/kishlaly/ta/analyze/testing/sl/StopLossVolatileATR.java
package com.kishlaly.ta.analyze.testing.sl;

import com.kishlaly.ta.model.QuoteJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.ATRJava;
import com.kishlaly.ta.utils.IndicatorUtilsJava;

import java.util.List;

/**
 * SL = Current low – (2 × ATR)
 */
public class StopLossVolatileATR extends StopLossStrategyJava {

    public StopLossVolatileATR() {
        super(null, true);
    }

    @Override
    public double calculate(SymbolDataJava data, int currentQuoteIndex) {
        QuoteJava signal = data.quotes.get(currentQuoteIndex);
        List<ATRJava> atrs = IndicatorUtilsJava.buildATR(data.symbol, data.quotes, 22);
        return signal.getLow() - (2 * atrs.get(currentQuoteIndex).getValue());
    }

    @Override
    public String toString() {
        return "SL volatile ATR";
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/testing/sl/StopLossVolatileATR.kt
package com.kishlaly.ta.analyze.testing.sl

import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.utils.IndicatorUtils

/**
 * SL = Current low – (2 × ATR)
 */
class StopLossVolatileATR : StopLossStrategy(isVolatile = true) {

    override fun calculate(data: SymbolData, signalIndex: Int): Double {
        val signal = data.quotes[signalIndex]
        val atrs = IndicatorUtils.buildATR(data.symbol, data.quotes, 22)
        return signal.low - 2 * atrs[signalIndex].value
    }

    override fun toString() = "SL volatile ATR"

}